Update the mscorlib.md document. (#36225)
dotnet/runtime: docs/design/coreclr/botr/mscorlib.md

> Mscorlib (a.k.a. `System.Private.CoreLib`) and calling into the Runtime

Reviewer: Why not eliminate the mscorlib term entirely (and rename the doc)? Relegate it to a brief historical mention.

Reviewer: Ha, now I read your changes. But I still suggest renaming the doc.

Author: Not against that. If I was going to do that, I would probably also rename `mscorlib.h` and `mscorlib.cpp` as well. Not in this PR, but in a subsequent one. What is the appetite for excising the term "mscorlib" from the functional code base? Is that a longer term goal?

Reviewer: We did rename the folder it was in, and no doubt some identifiers/comments. It does not seem worth special or destabilizing work such as changing `mscorlib.h` given all the other inconsistencies and irregularities in the VM code. But, that is just my 2c.

Reviewer 2: We have done several passes of the mscorlib -> CoreLib renaming. If you want to rename what fell through the cracks, it is fine with me.

Author: Will change the name of the file.

===

# Introduction

> Mscorlib is the assembly for defining the core parts of the type system, and a good portion of the Base Class Library in .NET Framework. It has been renamed to `System.Private.CoreLib` (SPCL) in .NET Core, though many places in the code and documentation still refer to it as mscorlib. This document will endevour to stick to using `System.Private.CoreLib` or SPCL. Base data types live in this assembly, and it has a tight coupling with the CLR. Here you will learn exactly how and why SPCL is special, and the basics about calling into the CLR from managed code via QCall and FCall methods. It also discusses calling from within the CLR into managed code.

Reviewer: ```suggestion
Mscorlib is the assembly for defining the core parts of the type system, and a good portion of the Base Class Library in .NET Framework. It has been renamed to `System.Private.CoreLib` (SPCL) in .NET Core, though many places in the code and documentation still refer to it as mscorlib. This document will endeavour to stick to using `System.Private.CoreLib` or SPCL. Base data types live in this assembly, and it has a tight coupling with the CLR. Here you will learn exactly how and why SPCL is special, and the basics about calling into the CLR from managed code via QCall and FCall methods. It also discusses calling from within the CLR into managed code.
```

## Dependencies

> Since SPCL defines base data types like `Object`, `Int32`, and `String`, SPCL cannot depend on other managed assemblies. However, there is a strong dependency between SPCL and the CLR. Many of the types in SPCL need to be accessed from native code, so the layout of many managed types is defined both in managed code and in native code inside the CLR. Additionally, some fields may be defined only in Debug, Checked, or Release builds, so typically SPCL must be compiled separately for each type of build.

Reviewer 2: I would use `CoreLib` as the short name. I think it is easier to understand and we are using that name much more frequently today.

E.g. I can see 2 instances of `SPCL` under CoreCLR, but several dozens of `CoreLib`.

Reviewer 2: Out of curiosity, I have looked up possible SPCL meanings at https://acronyms.thefreedictionary.com/SPCL . From now on, we should call it "Synthetic Peptide Combinatorial Library" :-)

Author: Definitely changing to "CoreLib"...

> For 64 bit platforms, some constants are also defined at compile time. So a 64 bit `System.Private.CoreLib.dll` is slightly different from a 32 bit `System.Private.CoreLib.dll`. Due to these constants, such as `IntPtr.Size`, most libraries above SPCL should not need to build separately for 32 bit vs. 64 bit.

Reviewer: ```suggestion
System.Private.CoreLib.dll builds separately for 64 bit and 32 bit, and some public constants it exposes differ by bitness. By using these constants, such as `IntPtr.Size`, most libraries above SPCL should not need to build separately for 32 bit vs. 64 bit.
```

# Calling from managed to native code

Two techniques exist for calling into the CLR from managed code. FCall allows you to call directly into the CLR code, and provides a lot of flexibility in terms of manipulating objects, though it is easy to cause GC holes by not tracking object references correctly. QCall also allows you to call into the CLR via the P/Invoke, but is much harder to accidentally mis-use. FCalls are identified in managed code as extern methods with the [`MethodImplOptions.InternalCall`](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.methodimploptions) bit set. QCalls are marked `static extern` methods similar to regular P/Invokes, but are directed toward a library called `"QCall"`.

There is a small variant of FCall called HCall (for Helper call) for implementing JIT helpers. The HCall is intended for doing things like accessing multi-dimensional array elements, range checks, etc. The only difference between HCall and FCall is that HCall methods won't show up in an exception stack trace.

### Choosing between FCall, QCall, P/Invoke, and writing in managed code

> First, remember that you should be writing as much as possible in managed code. You avoid a raft of potential GC hole issues, you get a better debugging experience, and the code is often simpler. It also continues the ongoing refactoring of SPCL into smaller layered fully [managed libraries](https://github.com/dotnet/runtime/src/libraries).

Reviewer 2: "refactoring of SPCL into smaller layered fully libraries" is not a think anymore. You can delete it.

> Reasons to write FCalls in the past generally fell into three camps: missing language features, better performance, or implementing unique interactions with the runtime. C# now has almost every useful language feature that you could get from C++, including unsafe code and stack-allocated buffers, and this eliminates the first two reasons for FCalls. We have ported some parts of the CLR that were heavily reliant on FCalls to managed code in the past (such as Reflection, some Encoding, and String operations) and we intend to continue this momentum.

Reviewer: Do we have a meta issue tracking components that we would like to port to managed? I am not sure we have one, but it feels like it would be worth having. @jkotas.

Reviewer 2: Yes, we can have one. 

There are many small ones like https://github.com/dotnet/runtime/pull/33701 or https://github.com/dotnet/runtime/pull/32722 done recently.

Out of the big ones:

- The top of the list is threadpool that is being worked on
- The next one with the best cost/benefit is probably Reflection.Emit (https://github.com/dotnet/runtime/issues/15704)

If the only reason you're defining a FCall method is to call a native method, you should be using P/Invoke to call the method directly. [P/Invoke](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.dllimportattribute) is the public native method interface and should be doing everything you need in a correct manner.

If you still need to implement a feature inside the runtime, consider if there is a way to reduce the frequency of transitioning to native code. Can you write the common case in managed and only call into native for some rare corner cases? You're usually best off keeping as much as possible in managed code.

QCalls are the preferred mechanism going forward. You should only use FCalls when you are "forced" to. This happens when there is common "short path" through the code that is important to optimize. This short path should not be more than a few hundred instructions, cannot allocate GC memory, take locks or throw exceptions (`GC_NOTRIGGER`, `NOTHROWS`). In all other circumstances (and especially when you enter a FCall and then simply erect HelperMethodFrame), you should be using QCall.

> FCalls were specifically designed for short paths of code that must be optimized. They allowed explicit control over when erecting a frame was done. However, it is error prone and not worth the complexity for many APIs. QCalls are essentially P/Invokes into the CLR. In the event the performance of an FCall is required consider creating a QCall and marking it with [`SuppressGCTransitionAttribute`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.suppressgctransitionattribute).

Reviewer 3: Not necessarily in this doc but something to think about: The normal `QCALL_CONTRACT` and `QCALL_CHECK` macros specify the GC mode so they can't be used on QCalls that use `SuppressGCTransition`. We should add macros for "suppressed-gc-transition QCall" contracts/checks to make it easier for developers to define them correctly.

Author: Agreed. Can you add an issue for that?

Reviewer 3: Opened https://github.com/dotnet/runtime/issues/36238

Failing to properly report an `OBJECTREF` or to update an interior pointer is commonly referred to as a "GC hole", because the `OBJECTREF` class will do some validation that it points to a valid object every time you dereference it in Debug and Checked builds. When an `OBJECTREF` pointing to an invalid object is dereferenced, an assert will trigger saying something like "Detected an invalid object reference. Possible GC hole?". This assert is unfortunately easy to hit when writing "manually managed" code.

Note that QCall's programming model is restrictive to sidestep GC holes by forcing you to pass in the address of an object reference on the stack. This guarantees that the object reference is GC protected by the JIT's reporting logic, and that the actual object reference will not move because it is not allocated in the GC heap. QCall is our recommended approach, precisely because it makes GC holes harder to write.

> ### FCall epilogue walker for x86

Reviewer: ```suggestion
### FCall epilog walker for x86
```
I'm a British English speaker, but you use epilog everywhere else.

    {
        if (str == null)
        {
            throw new ArgumentNullException(nameof(str));
        }

        return str.IsInterned();
    }
}
```

### FCall example – unmanaged

The FCall entrypoint has to be registered in tables in [vm\ecalllist.h][ecalllist] using `FCFuncEntry` macro. See ["Registering your QCall or FCall Method"](#register).

This method is an instance method in managed code, with the "this" parameter passed as the first argument. We use `StringObject*` as the argument type, then copy it into a `STRINGREF` so we get some error checking when we use it.

```C++
FCIMPL1(Object*, AppDomainNative::IsStringInterned, StringObject* pStringUNSAFE)
{
    FCALL_CONTRACT;

    STRINGREF       refString   = ObjectToSTRINGREF(pStringUNSAFE);
    STRINGREF*      prefRetVal  = NULL;

    HELPER_METHOD_FRAME_BEGIN_RET_1(refString);

    if (refString == NULL)
        COMPlusThrow(kArgumentNullException, W("ArgumentNull_String"));

    prefRetVal = GetAppDomain()->IsStringInterned(&refString);

    HELPER_METHOD_FRAME_END();

    if (prefRetVal == NULL)
        return NULL;

    return OBJECTREFToObject(*prefRetVal);
}
FCIMPLEND
```

<a name="register"></a>

> ## Registering your QCall or FCall method

Reviewer 2: ```suggestion
## <a name="register" /> Registering your QCall or FCall method
```

I believe that this is the usual style used for this. Otherwise, it looks like part of previous paragraph when reading the raw text.

Here's a simplified example. Note how this instance uses the binder described in the previous section to call `SafeHandle`'s virtual `ReleaseHandle` method.

```C++
void SafeHandle::RunReleaseMethod(SafeHandle* psh)
{
    CONTRACTL {
        THROWS;
        GC_TRIGGERS;
        MODE_COOPERATIVE;
    } CONTRACTL_END;

    SAFEHANDLEREF sh(psh);

    GCPROTECT_BEGIN(sh);

    MethodDescCallSite releaseHandle(s_pReleaseHandleMethod, METHOD__SAFE_HANDLE__RELEASE_HANDLE, (OBJECTREF*)&sh, TypeHandle(), TRUE);

    ARG_SLOT releaseArgs[] = { ObjToArgSlot(sh) };
    if (!(BOOL)releaseHandle.Call_RetBool(releaseArgs)) {
        MDA_TRIGGER_ASSISTANT(ReleaseHandleFailed, ReportViolation)(sh->GetTypeHandle(), sh->m_handle);
    }

    GCPROTECT_END();
}
```

# Interactions with Other Subsystems

## Debugger

One limitation of FCalls today is that you cannot easily debug both managed code and FCalls easily in Visual Studio's Interop (or mixed mode) debugging. Setting a breakpoint today in an FCall and debugging with Interop debugging just doesn't work. This most likely won't be fixed.

> # Physical architecture

Reviewer 2: I see that you have lower-cases some titles (except the first word), but not all of them. Is there a pattern or is it just an oversight?

Author: Probably an oversight. Will go through again.
